quixstreams/sinks/community/iceberg.py

```diff
@@ -0,0 +1,203 @@
+    from pyiceberg.types import StringType, TimestampType
+    from pyiceberg.exceptions import CommitFailedException  # Import the exception
+except ImportError as exc:
+    raise ImportError(
+        'Package "pyiceberg" is missing: '
+        "run pip install quixstreams[pyiceberg] to fix it"
+    ) from exc
+
+
+DataCatalogSpec = Literal["aws_glue"]
+
+LogLevel = Literal[
+    "CRITICAL",
+    "ERROR",
+    "WARNING",
+    "INFO",
+    "DEBUG",
+    "NOTSET",
+]
+
+class IcebergSink(BatchingSink):
+    """
+    IcebergSink is a sink that writes batches of data to an Apache Iceberg table stored in AWS S3,
+    using the AWS Glue Data Catalog as a default catalog (only implemented for now).
+
+    It serializes incoming data batches into Parquet format and appends them to the Iceberg table,
+    updating the table schema as necessary on fly.
+    """
+    def __init__(
+        self,
+        table_name: str,
+        aws_s3_uri: str,
+        s3_region_name: Optional[str] = None,
+        data_catalog_spec: DataCatalogSpec = "aws_glue",
+        schema: Optional[Schema] = None,
+        partition_spec: Optional[PartitionSpec] = None,
+        loglevel: LogLevel = "INFO"
+    ):
+        """
+        Initializes the S3Sink with the specified configuration.
+
+        Parameters:
+            table_name (str): The name of the Iceberg table.
+            aws_s3_uri (str): The S3 URI where the table data will be stored (e.g., 's3://your-bucket/warehouse/').
+            s3_region_name (Optional[str]): The AWS region where the S3 bucket and Glue catalog are located.
+            data_catalog_spec (DataCatalogSpec): The data catalog specification to use (default is 'aws_glue').
+            schema (Optional[Schema]): The Iceberg table schema. If None, a default schema is used.
+            partition_spec (Optional[PartitionSpec]): The partition specification for the table. If None, a default is used.
+            format: The data serialization format to use (default is ParquetFormat()).
+            loglevel (LogLevel): The logging level for the logger (default is 'INFO').
+        """
+        super().__init__()
+        
+        # Configure logging.
+        self._logger = logging.getLogger("IcebergSink")
+        log_format = '[%(asctime)s.%(msecs)03d] [%(levelname)s] [%(name)s] : %(message)s'
+        logging.basicConfig(format=log_format, datefmt='%Y-%m-%d %H:%M:%S')
+        self._logger.setLevel(loglevel)
+
+        # Initialize the Iceberg catalog.
+        if data_catalog_spec == "aws_glue":
+            # Configure Iceberg Catalog using AWS Glue.
+            self.catalog = GlueCatalog(
+                name="glue_catalog",
+                region_name=s3_region_name
+            )
+        else:
+            raise ValueError(f"Unsupported data_catalog_spec: {data_catalog_spec}")
+        
+        # Set up the schema.
+        if schema is None:
+            # Define a default schema if none is provided.
+            schema = Schema(
```

Comment:
Don't know enough about Iceberg, but is it "safe" or sensible to have/use a default schema?

Reply:
yes, we always have timestamp and key